interfaces/apparmor/backend.go

```diff
@@ -285,6 +286,31 @@ func profileGlobs(snapName string) []string {
 	return []string{interfaces.SecurityTagGlob(snapName), nsProfile(snapName)}
 }
 
+// Determine if a profile filename is removable during core refresh/rollback.
+// This is needed because core devices are also special, the apparmor cache
+// gets confused too easy, especially at rollbacks, so we delete the cache. See
+// Setup(), below. Some systems employ a unified cache directory where all
+// apparmor cache files are stored under one location so ensure we don't remove
+// the snap profiles since snapd manages them elsewhere and instead only remove
+// snap-confine and system profiles (eg, as shipped by distro package manager
+// or created by the administrator). snap-confine profiles are like the
+// following:
+// - usr.lib.snapd.snap-confine.real
+// - usr.lib.snapd.snap-confine
+// - snap.core.NNNN.usr.lib.snapd.snap-confine
+// - var.lib.snapd.snap.core.NNNN.usr.lib.snapd.snap-confine
+// - snap-confine.core.NNNN
+// TODO: also the "snapd" snap here soon
+func profileIsRemovableOnCoreSetup(fn string) bool {
+	bn := path.Base(fn)
+	if strings.HasPrefix(bn, ".") {
+		return false
+	} else if strings.HasPrefix(bn, "snap") && !strings.HasPrefix(bn, "snap-confine.core.") && !strings.Contains(bn, "usr.lib.snapd.snap-confine") {
```

Comment:
Should the first test check for `snap.`?

Reply:
I think "snap" is covering both "snap." and "snap-update-ns.", maybe it should be more explicit and use both

Reply:
They idea is that we want to skip all policy for the snaps themselves. There are a bunch of ways to do this and I picked what I deemed to be the simplest to understand and was future-proofed and picked the idea of "skip everything that starts with 'snap' since snapd manages that, except for snap-confine which we want to regenerate".

Reply:
There could be something called snapshot... for example  that would also match though?

Reply:
@pedronis - yes, it would match and the profile would not be regenerated for this profile. Considering that this only fires on core I opted for simpler to understand and didn't worry about non-snapd things that might start with 'snap*' because those would only come from the distro and the distro doesn't have anything in the core/snapd snap (and hard to imagine ever since it is so small) that would start with snap that doesn't come from snapd. I'll also mention that one of the reasons I didn't break this out in the first place and had it down in the workaround code was so that this point was more clear. I was asked to break the check into a function, which I did, but it makes it harder to understand the context it is operating under imho.

Reply:
mmh, the question is a bit what is more likely:
* the something with snap... get added to core
* that we add some more profiles related to snaps  but forget to change this bit

Reply:
@pedronis - sure. The latter is the one that is more possible which then you must ask which is more likely:
* the profiles that are added are something we don't want removed (ie, they are like snap.*, snap-update-ns.*)
* the profiles that are added are something we want removed (ie, they are like snap-confine)

I felt the former was more likely and so coded it as such. Problem is, if you code it one way and add the other at a later date, then you did the wrong thing.